File: CPP/learn_cpp/codewars/valid_braces.cpp
#include <vector>
include <string>
include <iostream>
include <vector>
void main() {
    std::cout << valid_braces("(){}[]") << "\n";
}

bool valid_braces(const std::string &braces) {
    std::vector<char> b = {};
    for (char c: braces){
        if (b.size() == 0) b.push_back(c);
        else {
            for(int i = 0; i < b.size(); i++) {
                char bb{b[i]};
                int end = b.size();
                if (is_match(c, bb)){
                    char temp = bb;
                    b[i] = b[end];
                    b[end] = temp;
                    b.pop_back();
                    break;
                }

            }
        }
    }
    if (b.size() == 0) return true;
    return false;
}


bool is_match(char b1, char b2){
    if (b1 == '{' && b2 == '}'){
        return true;
    }
    else if (b1 == '(' && b2 == ')'){
        return true;
    }
    else if (b1 == '[' && b2 == ']') {
        return true;
    }
    return false;
}

File: CPP_n_C/learn_cpp/codewars/valid_braces.py

def valid_braces(braces:str):
    b = []
    for c in braces:
        if len(b) == 0: b.append(c)
        else:
            for bb in b:
                if is_match(c, bb):
                    b.remove(bb)
                    break
    if len(b) == 0: return True
    return False
                        

def is_match(b1, b2):
    if b1 == '{' and b2 == '}':
        return True
    elif b1 == '(' and b2 == ')':
        return True
    elif b1 == '[' and b2 == ']':
        return True
    return False

def main():
    print(valid_braces("{[()]}"))

main()
